src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+  mDialog->show();
+}
+
+void Qgs3DMapToolMeasureLine::deactivate()
+{
+  qInfo() << "Measure line deactivated";
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    disconnect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+
+  // Delete previouse line
+  mMeasurementLayer->startEditing();
+  mMeasurementLayer->deleteFeature( mMeasurementFeature->id() );
+  mMeasurementLayer->commitChanges();
+  mMeasurementLine->clear();
+
+  mCanvas->map()->setRenderers( QList<QgsAbstract3DRenderer *>() << mMeasurementLayer->renderer3D()->clone() );
+
+  mCanvas->scene()->unregisterPickHandler( mPickHandler.get() );
+
+  // Hide dialog
+  mDialog->hide();
+}
+
+void Qgs3DMapToolMeasureLine::onTerrainPicked( Qt3DRender::QPickEvent *event )
+{
+  qInfo() << "onTerrainPicked";
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mDone )
+    {
+      mDialog->restart();
+    }
+
+    const QVector3D worldIntersection = event->worldIntersection();
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates( QgsVector3D( worldIntersection.x(),
+                            worldIntersection.y(),
+                            worldIntersection.z() ), mCanvas->map()->origin() );
+    qInfo() << "Coord (onTerrainPicked): " << mapCoords.x() << " " << mapCoords.y() << " " << mapCoords.z();
+
+    mDone = false;
+    addPoint( QgsPoint( mapCoords.x(), mapCoords.y(), mapCoords.z() ) );
+    mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    undo();
+  }
+
+
+
+}
+
+
+void Qgs3DMapToolMeasureLine::onTerrainEntityChanged()
+{
+  qInfo() << "onTerrainEntityChanged";
+  // no need to disconnect from the previous entity: it has been destroyed
+  // start listening to the new terrain entity
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+}
+
+void Qgs3DMapToolMeasureLine::setMeasurementLayerRenderer( QgsVectorLayer *layer )
+{
+  layer->setRenderer3D( mLineSymbolRenderer );
+  mLineSymbolRenderer->setLayer( layer );
+}
+
+
+void Qgs3DMapToolMeasureLine::addPoint( const QgsPoint &point )
+{
+  // don't add points with the same coordinates
+  if ( !mPoints.isEmpty() && mPoints.last() == point )
+  {
+    return;
+  }
+
+  QgsPoint addedPoint( point );
+  qInfo() << "Is 3D point? " << addedPoint.is3D();
+
+  // Append point
+  mPoints.append( addedPoint );
+
+  mMeasurementLine->addVertex( addedPoint );
+
+  mMeasurementLayer->startEditing();
+  QgsGeometry *newMeasurementLine = new QgsGeometry( mMeasurementLine );
```

Comment:
a simpler solution than modifying mMeasurementLine would be to just create a new QgsLineString from mPoints whenever we need to call changeGeometry() - like that we only keep the list of points in one place.

Reply:
done.